src/Events/Span.php

```diff
@@ -56,6 +56,7 @@ public function __construct(string $name, EventBean $parent)
     {
         parent::__construct([]);
         $this->name  = trim($name);
+        // TODO refactor to have a TimerFactory or something to create different timer implementations
```

Comment:
How allowing different implementations for `Timer` could be better in this case? `Timer` is a start/stop model and I can't see any good reason to add high abstractions here.

Reply:
My biggest is concern is testability. Since the Span class creates a Timer, there is not opportunity to inject a mock. I previously had to play some games in the AgentTest because of timing and it still fails occasionally just because of timing. Using a mock would enable much better testing. 

I can also see that we will want to implement a Timer using the hrtime() function introduced in PHP 7.3. Using an abstraction would let us use an implementation appropriate for the PHP version.

Reply:
> My biggest is concern is testability. Since the Span class creates a Timer, there is not opportunity to inject a mock. I previously had to play some games in the AgentTest because of timing and it still fails occasionally just because of timing. Using a mock would enable much better testing.

We can solve this by creating a protected factory method to create a new instance of Timer if it doesnt yet exist and create a mock class extending Span which overwrites that method.

```php
class Span {
    protected function createTimer() {...}
}

$mock = new class extends Span {
    protect function createTimer () {}
}
```

I'm always prefer abstractions but honestly changing Timer implementation here is not a common requirement for Spans IMHO. But having it here will not brings any high drawbacks, unless more code to maintain. 😉

Reply:
That seems reasonable and still fits with the "or something" in the TODO. :)

Reply:
I took a stab at implementing this in `SpanTest`. It works, but feels convoluted to me.

https://github.com/nipwaayoni/elastic-apm-php-agent/blob/event-timer-changes/tests/Events/SpanTest.php

Is there something I'm missing? 

I can assert that the extended `createTimer()` method is called with the expected start time. And I can assert that the mock `Timer` object it returns has the the `stop()` method called. But those to assertions are widely decoupled in the test and it takes some work to mentally follow the test setup and execution. I'll look for ways to clean that up, but I wanted to get a working implementation out first.

Reply:
Comments on up to dated source code.